refactor Innerjoin

zengyan1 opened the pull request:
### **User description**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:


___

### **PR Type**
Enhancement


___

### **Description**
- Refactored `InnerJoin` struct to use a container struct instead of a pointer.
- Optimized initialization in the `Prepare` method to avoid redundant allocations.
- Updated `Call` and `build` methods to use a pointer to the container for consistency.
- Improved memory management in `Reset` and `Free` methods by cleaning up resources more effectively.
- Enhanced `cleanBatch` method to ensure proper batch cleanup.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>join.go</strong><dd><code>Refactor InnerJoin methods for optimized initialization and pointer </code><br><code>usage</code></dd></summary>
<hr>

pkg/sql/colexec/join/join.go

<li>Refactored <code>Prepare</code> method to initialize vectors and executors only if <br>they are nil.<br> <li> Modified <code>Call</code> method to use a pointer to <code>innerJoin.ctr</code>.<br> <li> Adjusted <code>build</code> method to use a pointer to <code>innerJoin.ctr</code>.<br>


</details>


  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17927/files#diff-2a43ca762c59d0ece91404da915aa519c35db896df1ec228aab36a10e7dbc1dd">+15/-11</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>types.go</strong><dd><code>Refactor InnerJoin struct and methods for improved memory management</code></dd></summary>
<hr>

pkg/sql/colexec/join/types.go

<li>Changed <code>ctr</code> from a pointer to a struct in <code>InnerJoin</code>.<br> <li> Refactored <code>Reset</code> and <code>Free</code> methods to use a pointer to <code>innerJoin.ctr</code>.<br> <li> Enhanced <code>cleanBatch</code> method to handle batch cleanup properly.<br>


</details>


  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17927/files#diff-67c85cf1b10e35ddf7ac938c17226a987e5548785b126d8c3929a26874fa1d78">+21/-18</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

badboynt1 commented on pkg/sql/colexec/join/types.go:
```diff
@@ -149,6 +148,10 @@ func (ctr *container) cleanExprExecutor() {
+		ctr.bat = nil
+	}
 	if ctr.rbat != nil {
 		proc.PutBatch(ctr.rbat)
```
所有的putbatch都要删掉

zengyan1 replied:
已删除

badboynt1 commented on pkg/sql/colexec/join/types.go:
```diff
@@ -113,29 +113,28 @@ func (innerJoin *InnerJoin) Release() {
+	ctr := &innerJoin.ctr
+	anal := proc.GetAnalyze(innerJoin.GetIdx(), innerJoin.GetParallelIdx(), innerJoin.GetParallelMajor())
+
+	ctr.cleanBatch(proc)
```
reset函数里应该不用cleanbatch  因为下次重新执行是可以复用的

ouyuanning replied:
那3个batch的生命周期我不太了解，如果是一直存在的，看prepare甚至new算子的时候是否可以先new出来？

zengyan1 replied:
已修改

fengttt commented on pkg/sql/colexec/join/types.go:
```diff
@@ -149,6 +148,10 @@ func (ctr *container) cleanExprExecutor() {
 func (ctr *container) cleanBatch(proc *process.Process) {
 	ctr.batches = nil
+	if ctr.bat != nil {
+		proc.PutBatch(ctr.bat)
```
Said, nobody can use PutBatch/GetBatch anymore.

zengyan1 replied:
已删除

fengttt reviewed (changes requested):
So you missed the point of this change.

1. You are NOT allowed to use PutBatch and GetBatch,
2. Hold the batch in your ctr, as member.
3. Make sure you understand life span of batches.   
4.

ouyuanning commented on pkg/sql/colexec/join/types.go:
```diff
@@ -113,29 +113,28 @@ func (innerJoin *InnerJoin) Release() {
+		innerJoin.ResetProjection(proc)
+	} else {
+		anal.Alloc(innerJoin.ctr.maxAllocSize)
+	}
```
innerJoin.ctr.maxAllocSize reset的时候要充值为0

ouyuanning commented:
UT补一下

fengttt commented on pkg/sql/colexec/join/join.go:
```diff
@@ -116,19 +115,16 @@ func (innerJoin *InnerJoin) Call(proc *process.Process) (vm.CallResult, error) {
-				if err != nil {
-					return result, err
-				}
+			result.Batch, err = innerJoin.EvalProjection(result.Batch, proc)
```
Can you do this?   You destroyed result.Batch,

zengyan1 replied:
result.Batch只是一个指针，其内容在ctr.rbat中

fengttt commented on pkg/sql/colexec/join/join.go:
```diff
@@ -116,19 +115,16 @@ func (innerJoin *InnerJoin) Call(proc *process.Process) (vm.CallResult, error) {
-				}
+			result.Batch, err = innerJoin.EvalProjection(result.Batch, proc)
+			if err != nil {
+				return result, err
```
Let's not reuse result.   Create more confusion than necessary.   Why don't you name the result from child's call `input` and use `result` for output of the operator?

zengyan1 replied:
已修改

fengttt commented on pkg/sql/colexec/join/join.go:
```diff
@@ -155,18 +151,24 @@ func (innerJoin *InnerJoin) build(anal process.Analyze, proc *process.Process) {
+		ctr.rbat = batch.NewWithSize(len(ap.Result))
+		for i, rp := range ap.Result {
+			if rp.Rel == 0 {
+				ctr.rbat.Vecs[i] = proc.GetVector(*ap.ctr.bat.Vecs[rp.Pos].GetType())
```
You will not use GetVector either.   Remove that crap.

zengyan1 replied:
已修改

fengttt commented on pkg/sql/colexec/join/types.go:
```diff
@@ -113,31 +108,43 @@ func (innerJoin *InnerJoin) Release() {
+	ctr.resetExprExecutor()
+	ctr.cleanHashMap() // -> resetHashMap
+	ctr.bat = nil      // children[0] clean
+	ctr.batches = nil  // reset/free in hashmap
```
About this one.   Too.

zengyan1 replied:
同下

fengttt commented on pkg/sql/colexec/join/types.go:
```diff
@@ -113,31 +108,43 @@ func (innerJoin *InnerJoin) Release() {
+	ctr.resetExecutor()
+	ctr.resetExprExecutor()
+	ctr.cleanHashMap() // -> resetHashMap
+	ctr.bat = nil      // children[0] clean
```
Bad idea -- why you want to release ctr.bat?   Is ctr.bat created in join?

zengyan1 replied:
这里没有release，只是为下次execute给一个nil的初值，ctr.bat和ctr.batches不在join中创建

zengyan1 commented:
以下是对ctr中几个bat的解释，已添加在join/types.go中 
```
1. batches means the data of right table. It's created by HashBuild operator and cleaned by cleanHashMap() in InnerJoin.Reset().
2. inbat means the data of left table. It's created by InnerJoin.Children[0] and cleaned by InnerJoin.Children[0].
3. rbat means the result of InnerJoin. InnerJoin.Probe() create rbat once when it's called first time and use CleanOnlyData() after that.
   InnerJoin.Reset() doesn't need to reset or clean rbat, because the result always has same types. InnerJoin.Free() will clean rbat.
4. joinBat1 means some data from left table used to evaluate join conditions. InnerJoin.Probe() create joinBat1 once.
   joinBat1 will always be overwritten by colexec.SetJoinBatchValues().
   InnerJoin.Reset() doesn't need to reset or clean joinBat1. InnerJoin.Free() will clean joinBat1.
5. joinBat2 means some data from right table, same as joinBat1.
```

fengttt commented on pkg/sql/colexec/join/types.go:
```diff
@@ -55,17 +51,28 @@ type container struct {
 
+/*
+InnerJoin.container has 5 *batch or []*batch
+1. batches means the data of right table. It's created by HashBuild operator and cleaned by cleanHashMap() in InnerJoin.Reset().
```
If it is created by HashBuild, why do we need to clean in InnerJoin.Reset()?   Should HashBuild clean it?

badboynt1 replied:
hashbuild不能clean这部分数据，因为probe的时候还要访问。 这部分数据和hash表一起封装到joinmapMsg里面，通过引用计数来管理，引用计数为0时才能释放。

zengyan1 commented:
已删除ctr.batches，改为从joinmap直接访问

ouyuanning commented on pkg/vm/message/joinMapMsg.go:
```diff
@@ -54,7 +54,7 @@ func (jm *JoinMap) GetBatches() []*batch.Batch {
 		return nil
 	}
-	return jm.batches
+	return jm.Batches
```
已经有GetBatches() 方法了，没必要修改这里的batches成public的啊

zengyan1 replied:
已修改
